src/excalidraw-app/index.tsx

```diff
@@ -9,6 +9,131 @@ import { importFromLocalStorage } from "../data/localStorage";
+import { loadFromFirebase } from "./data/firebase";
+import { restore } from "../data/restore";
+
+const shouldForceLoadScene = (
+  scene: ResolutionType<typeof loadScene>,
+): boolean => {
+  if (!scene.elements.length) {
+    return true;
+  }
+
+  const roomMatch = getCollaborationLinkData(window.location.href);
+
+  if (!roomMatch) {
+    return false;
+  }
+
+  const roomID = roomMatch[1];
+
+  let collabForceLoadFlag;
+  try {
+    collabForceLoadFlag = localStorage?.getItem(
+      LOCAL_STORAGE_KEY_COLLAB_FORCE_FLAG,
+    );
+  } catch {}
+
+  if (collabForceLoadFlag) {
+    try {
+      const {
+        room: previousRoom,
+        timestamp,
+      }: { room: string; timestamp: number } = JSON.parse(collabForceLoadFlag);
+      // if loading same room as the one previously unloaded within 15sec
+      //  force reload without prompting
+      if (previousRoom === roomID && Date.now() - timestamp < 15000) {
+        return true;
+      }
+    } catch {}
+  }
+  return false;
+};
+
+type Scene = ResolutionType<typeof loadScene>;
+const initializeScene = async (opts?: {
+  onLateInitialization?: (data: { scene: Scene }) => void;
+}): Promise<Scene | false> => {
+  const searchParams = new URLSearchParams(window.location.search);
+  const id = searchParams.get("id");
+  const jsonMatch = window.location.hash.match(
+    /^#json=([0-9]+),([a-zA-Z0-9_-]+)$/,
+  );
+
+  const initialData = importFromLocalStorage();
+
+  let scene = await loadScene(null, null, initialData);
+
+  let isCollabScene = !!getCollaborationLinkData(window.location.href);
+  const isExternalScene = !!(id || jsonMatch || isCollabScene);
+  if (isExternalScene) {
+    if (
+      shouldForceLoadScene(scene) ||
+      window.confirm(t("alerts.loadSceneOverridePrompt"))
+    ) {
+      // Backwards compatibility with legacy url format
+      if (id) {
+        scene = await loadScene(id, null, initialData);
+      } else if (jsonMatch) {
+        scene = await loadScene(jsonMatch[1], jsonMatch[2], initialData);
+      }
+      if (!isCollabScene) {
+        window.history.replaceState({}, "Excalidraw", window.location.origin);
+      }
+    } else {
+      // https://github.com/excalidraw/excalidraw/issues/1919
+      if (document.hidden) {
+        window.addEventListener(
+          "focus",
+          () =>
+            initializeScene().then((_scene) => {
+              opts?.onLateInitialization?.({ scene: _scene || scene });
+            }),
+          {
+            once: true,
+          },
+        );
+        return false;
+      }
+
+      isCollabScene = false;
+      window.history.replaceState({}, "Excalidraw", window.location.origin);
+    }
+  }
+  if (isCollabScene) {
+    // TODO
+    // when joining a room we don't want user's local scene data to be merged
+    //  into the remote scene
+    // this.excalidrawRef.current.resetScene();
+    // await props.collab.initializeSocketClient({ showLoadingState: true });
```

Comment:
blocked on making collab socket initializer be able to be invoked lazily after the core component is initialized (and `excalidrawRef` is populated).

Reply:
since the child gets mounted first so ref should be accessible in the parent but here the order of mounting order seems to be reversed hence ref is not available

Reply:
It's not reversed, except that between `<ExcalidrawApp/>` and `<App/>` sits `<InitializeApp/>` that delays `<App/>` mount.

In general, we shouldn't depend on mounting order because that can change at any time.

Reply:
got it. Then one way to handle this would be to have an `afterLoad` callback which gets triggered once excalidraw is mounted and all initialization stuff should be done from inside that callback. That way we will always have ref accessible. WDYT?